scripts/check-auth.js

```diff
@@ -0,0 +1,137 @@
+  }
+  return { valid, errors };
+}
+
+walkSync("./contracts/").forEach(contractName => {
+  // Only contracts using domain-level permissions need to be checked, i.e. those that implement
+  // functions in IColony.sol
+  // Basically only Colony.sol, ColonyFunding.sol, and ColonyTask.sol (?)
+  if (
+    [
+      "contracts/ens/ENS.sol",
+      "contracts/ens/ENSRegistry.sol",
+      "contracts/gnosis/MultiSigWallet.sol",
+      "contracts/PatriciaTree/Bits.sol",
+      "contracts/PatriciaTree/Data.sol",
+      "contracts/PatriciaTree/IPatriciaTree.sol",
+      "contracts/PatriciaTree/IPatriciaTreeNoHash.sol",
+      "contracts/PatriciaTree/PatriciaTree.sol",
+      "contracts/PatriciaTree/PatriciaTreeNoHash.sol",
+      "contracts/PatriciaTree/PatriciaTreeBase.sol",
+      "contracts/PatriciaTree/PatriciaTreeProofs.sol",
+      "contracts/CommonAuthority.sol",
+      "contracts/ColonyAuthority.sol",
+      "contracts/ColonyNetwork.sol",
+      "contracts/ColonyNetworkAuction.sol",
+      "contracts/ColonyNetworkENS.sol",
+      "contracts/ColonyNetworkMining.sol",
+      "contracts/ColonyNetworkAuthority.sol",
+      "contracts/ColonyNetworkStorage.sol",
+      "contracts/ColonyStorage.sol",
+      "contracts/ContractEditing.sol",
+      "contracts/ERC20Extended.sol",
+      "contracts/EtherRouter.sol",
+      "contracts/IRecovery.sol",
+      "contracts/IColony.sol",
+      "contracts/IMetaColony.sol",
+      "contracts/IColonyNetwork.sol",
+      "contracts/IReputationMiningCycle.sol",
+      "contracts/ITokenLocking.sol",
+      "contracts/Migrations.sol",
+      "contracts/ReputationMiningCycle.sol",
+      "contracts/ReputationMiningCycleRespond.sol",
+      "contracts/Resolver.sol",
+      "contracts/TokenLocking.sol",
+      "contracts/TokenLockingStorage.sol",
+      "contracts/Token.sol", // Imported from colonyToken repo
+      "contracts/TokenAuthority.sol" // Imported from colonyToken repo
+    ].indexOf(contractName) > -1
+  ) {
+    return;
+  }
+
+  // Skip non-solidity files
+  if (contractName.indexOf(".sol") < 0) {
+    return;
+  }
+
+  const src = fs.readFileSync(`./${contractName}`, "utf8");
+
+  const result = parser.parse(src, { tolerant: true });
+  // Filters out an unknown number of 'pragmas' that we have.
+  const contract = result.children.filter(child => child.type === "ContractDefinition")[0];
+
+  // Check for that all public, non-{view,pure} functions have either stoppable or recovery modifiers.
+  contract.subNodes
+    .filter(child => child.type === "FunctionDefinition" && child.name !== "")
+    .forEach(functionDef => {
+      const res = correctAuthModifier(functionDef);
+      if (!res.valid) {
+        console.log(
+          "The contract",
+          contractName,
+          "doesn't appear to have the right auth declaration for function ",
+          functionDef.name,
+          ". Errors: ",
+          res.errors
+        );
+        process.exit(1);
+      }
+    });
+});
```

Comment:
Just as a side note here, we can consider decorating _**all**_ Colony functions with `auth` and explicitly marking open functions as ["public capability"](https://github.com/dapphub/ds-roles#setpubliccapability)

Reply:
We definitely could, but is it necessary? Seems like it would be more code for the same result.